### PR TITLE
fix: pass TError, TQueryData and TQueryKey to PlaceholderDataFunction

### PR DESCRIPTION
**Pull request by bviebahn**
The `previousData` and `previousQuery` parameters of the placeholder function seem to not be properly types. I think this should fix it. I tested it like this:
```typescript
const observer = new QueryObserver(
  {} as any,
  {
    queryKey: ['A', 'B'] as const,
    queryFn: () => 'Foo',
    placeholderData: (prev, prevQuery) => {
      // typeof key is now ['A', 'B'] instead of QueryKey
      const key = prevQuery?.queryKey
      return prev
    },
  },
)
```
Please let me know if I missed something.

### CODE REVIEW

**Comment by bviebahn** [comment]
Not sure what to do about the failing tests, running them locally they're all green. I only changed a type, so I think the failing test is not related to my change.

File: packages/query-core/src/types.ts [hunk 1]
```diff
@@ -319,7 +319,12 @@ export interface QueryObserverOptions<
    */
   placeholderData?:
     | NonFunctionGuard<TQueryData>
-    | PlaceholderDataFunction<NonFunctionGuard<TQueryData>>
+    | PlaceholderDataFunction<
+        NonFunctionGuard<TQueryData>,
```
Comment (bviebahn) [inline packages/query-core/src/types.ts:323]: I'm not sure if this should be `TQueryFnData`?

Comment (TkDodo) [reply packages/query-core/src/types.ts:323]: no I think this is fine. The difference is only for InfiniteQueries, where one is wrapped in `InfiniteData`. But maybe you can add a couple of type tests?

Comment (bviebahn) [reply packages/query-core/src/types.ts:323]: Sure, I added some tests. The last one would have also passed without my changes. Let me know if it needs more work.

**Review by TkDodo — APPROVED** [review]

**Comment by TkDodo** [comment]
format is failing

**Comment by bviebahn** [comment]
I don't understand why, when I run `pnpm run test:format` I'm not getting any errors. When I run `pnpm prettier --write` it gets formatted exactly as it currently is. Any idea?

**Comment by TkDodo** [comment]
just run the --write and commit it?

**Comment by TkDodo** [comment]
oh, likely you have the wrong version of prettier:

```
> prettier --plugin-search-dir . "{examples,integrations,packages,scripts}/**/*.{md,js,jsx,cjs,ts,tsx,json,vue,svelte}" "--check"

Checking formatting...
[warn] packages/query-core/src/types.ts
```

**Comment by bviebahn** [comment]
Oh, not sure what the problem was but after cloning the repo again I got the same prettier error. Should be fixed.

**Comment by TkDodo** [comment]
yeah actually, it was me changing something in a PR suggestion on a different PR, and merging it via the GH ui . sorry, it should be fixed on main now

**Pull request merged**